x-pack/legacy/plugins/uptime/public/state/api/utils.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import { isRight } from 'fp-ts/lib/Either';
+import { HttpFetchQuery, HttpSetup } from '../../../../../../../target/types/core/public';
+
+class ApiService {
+  private static instance: ApiService;
+  private _http!: HttpSetup;
+
+  public get http() {
+    return this._http;
+  }
+
+  public set http(httpSetup: HttpSetup) {
+    this._http = httpSetup;
+  }
+
+  private constructor() {}
+
+  static getInstance(): ApiService {
+    if (!ApiService.instance) {
+      ApiService.instance = new ApiService();
+    }
+
+    return ApiService.instance;
+  }
+
+  public async get(apiUrl: string, params?: HttpFetchQuery, decodeType?: any) {
```

Comment:
I think it would be better to require the `decodeType` parameter. We could then describe `decodeType`  better than `any`. If we did something like:

```ts
import { Decoder } from 'io-ts'
```

```suggestion
  public async get<T>(apiUrl: string, params?: HttpFetchQuery, decoder: Decoder<unknown, T>) {
```

In this case, `T` will be the type we get when `isRight` is truthy, and `unknown` is the type we specify for the value we give to `PathReporter` (which is ok, because `decode` will return an object of type `Either<t.Errors, T>` anyway, so the reporter will be happy).

`Partial`, `Type`, `Intersection`, etc., all seem to extend the `Decoder` interface, so I think this would be a safe assumption for us to make here.

Then, if we define a complex type like:
```typescript
const MyComplexType = t.intersection([
  t.partial({
    foo: t.string,
  }),
  t.type({
    bar: t.number,
  }),
]);
```

..the usage of the function will remain the same, but our responses will be typed appropriately:

```typescript
const apiResponse = apiService.get('/api/pathToComplexType', { foo: 'bar' }, MyComplexType);
```

And the resultant object will be typed like:
![image](https://user-images.githubusercontent.com/18429259/76860485-80188c00-6831-11ea-81be-c425d856d3b0.png)

If we allow the `decodeType` parameter to be optional, we can only ever guarantee a return type of `any`; this is more-or-less what we've always done, but the idea of type safety at runtime and "automatic" typing for all API responses sounds very nice to me.

If we feel it's out of the scope of this PR, a tech debt follow-up issue would be ok by me too, if we are agreed this is something we'd want.

Reply:
I created https://github.com/elastic/kibana/issues/60395 to track the addition of this. I think this will be a nice add when we can get to it.

Reply:
Thank you